File: src/python/ml_scratch/pure/perceptron.py
"""A dependency-free binary perceptron with deterministic initialization."""

from __future__ import annotations

import random
from collections.abc import Iterable, Sequence
from dataclasses import dataclass, field

from ml_scratch.pure.datasets import Dataset, Example, Features


@dataclass(frozen=True)
class TrainingResult:
    """Describe a completed training attempt without hiding non-convergence."""

    epochs: int
    converged: bool
    mistakes_per_epoch: tuple[int, ...]


@dataclass
class Perceptron:
    """Binary linear classifier trained with the classic perceptron update."""

    feature_count: int
    learning_rate: float = 0.1
    seed: int = 0
    weights: list[float] = field(init=False)
    bias: float = field(init=False)
    _random: random.Random = field(init=False, repr=False)

    def __post_init__(self) -> None:
        if self.feature_count <= 0:
            raise ValueError("feature_count must be positive")
        if self.learning_rate <= 0:
            raise ValueError("learning_rate must be positive")

        self._random = random.Random(self.seed)
        self.weights = [self._random.uniform(-0.5, 0.5) for _ in range(self.feature_count)]
        self.bias = self._random.uniform(-0.5, 0.5)

    def decision_score(self, features: Sequence[float]) -> float:
        """Return the signed score used by the step activation."""
        self._validate_features(features)
        return sum(weight * value for weight, value in zip(self.weights, features, strict=True)) + (
            self.bias
        )

    def predict(self, features: Sequence[float]) -> int:
        """Predict class 0 or 1 using a zero-threshold step activation."""
        return int(self.decision_score(features) >= 0.0)

    def predict_many(self, examples: Iterable[Features]) -> tuple[int, ...]:
        """Predict several feature vectors in their given order."""
        return tuple(self.predict(features) for features in examples)

    def fit(
        self, dataset: Dataset, *, max_epochs: int = 100, shuffle: bool = True
    ) -> TrainingResult:
        """Train until a mistake-free epoch or the epoch budget is exhausted."""
        examples = list(dataset)
        self._validate_training_data(examples, max_epochs)
        history: list[int] = []

        for epoch in range(1, max_epochs + 1):
            if shuffle:
                self._random.shuffle(examples)

            mistakes = 0
            for features, target in examples:
                error = target - self.predict(features)
                if error == 0:
                    continue
                adjustment = self.learning_rate * error
                self.weights = [
                    weight + adjustment * value
                    for weight, value in zip(self.weights, features, strict=True)
                ]
                self.bias += adjustment
                mistakes += 1

            history.append(mistakes)
            if mistakes == 0:
                return TrainingResult(epoch, True, tuple(history))

        return TrainingResult(max_epochs, False, tuple(history))

    def accuracy(self, dataset: Dataset) -> float:
        """Return the fraction of correctly classified examples."""
        examples = list(dataset)
        self._validate_training_data(examples, max_epochs=1)
        correct = sum(self.predict(features) == target for features, target in examples)
        return correct / len(examples)

    def _validate_features(self, features: Sequence[float]) -> None:
        if len(features) != self.feature_count:
            raise ValueError(f"expected {self.feature_count} features, received {len(features)}")

    def _validate_training_data(self, examples: list[Example], max_epochs: int) -> None:
        if max_epochs <= 0:
            raise ValueError("max_epochs must be positive")
        if not examples:
            raise ValueError("dataset must not be empty")
        for features, target in examples:
            self._validate_features(features)
            if target not in (0, 1):
                raise ValueError("targets must be binary values: 0 or 1")

File: src/cpp/src/perceptron.cpp
#include "ml_scratch/perceptron.hpp"

#include <algorithm>
#include <stdexcept>
#include <utility>

namespace ml_scratch {

BinaryPerceptron::BinaryPerceptron(const double learning_rate, const std::uint32_t seed)
    : learning_rate_{learning_rate}, random_engine_{seed} {
    if (learning_rate <= 0.0) {
        throw std::invalid_argument("learning_rate must be positive");
    }

    std::uniform_real_distribution<double> initial_value{-0.5, 0.5};
    for (double& weight : weights_) {
        weight = initial_value(random_engine_);
    }
    bias_ = initial_value(random_engine_);
}

double BinaryPerceptron::decision_score(const LogicInput& features) const noexcept {
    return weights_[0] * features[0] + weights_[1] * features[1] + bias_;
}

int BinaryPerceptron::predict(const LogicInput& features) const noexcept {
    return decision_score(features) >= 0.0 ? 1 : 0;
}

double BinaryPerceptron::accuracy(const LogicDataset& dataset) const noexcept {
    std::size_t correct = 0;
    for (const auto& [features, target] : dataset) {
        correct += predict(features) == target ? 1U : 0U;
    }
    return static_cast<double>(correct) / static_cast<double>(dataset.size());
}

PerceptronTrainingResult BinaryPerceptron::fit(LogicDataset dataset, const std::size_t max_epochs,
                                               const bool shuffle) {
    if (max_epochs == 0) {
        throw std::invalid_argument("max_epochs must be positive");
    }

    std::vector<std::size_t> history;
    history.reserve(max_epochs);

    for (std::size_t epoch = 1; epoch <= max_epochs; ++epoch) {
        if (shuffle) {
            std::shuffle(dataset.begin(), dataset.end(), random_engine_);
        }

        std::size_t mistakes = 0;
        for (const auto& [features, target] : dataset) {
            const int error = target - predict(features);
            if (error == 0) {
                continue;
            }

            const double adjustment = learning_rate_ * static_cast<double>(error);
            for (std::size_t index = 0; index < weights_.size(); ++index) {
                weights_[index] += adjustment * features[index];
            }
            bias_ += adjustment;
            ++mistakes;
        }

        history.push_back(mistakes);
        if (mistakes == 0) {
            return {epoch, true, std::move(history)};
        }
    }

    return {max_epochs, false, std::move(history)};
}

} // namespace ml_scratch
